core/drivers/clk/clk.c

```diff
@@ -7,19 +7,43 @@
 #include <config.h>
 #include <drivers/clk.h>
 #include <kernel/boot.h>
+#include <kernel/mutex.h>
 #include <kernel/panic.h>
-#include <kernel/spinlock.h>
+#include <kernel/thread.h>
 #include <malloc.h>
 #include <stddef.h>
 #include <stdio.h>
 
 /* Global clock tree lock */
-static unsigned int clk_lock = SPINLOCK_UNLOCK;
+static struct mutex clk_lock = MUTEX_INITIALIZER;
 
 #ifdef CFG_DRIVERS_CLK_PRINT_TREE
 static SLIST_HEAD(, clk) clock_list = SLIST_HEAD_INITIALIZER(clock_list);
 #endif
 
+static void lock_clk_access(void)
+{
+	/*
+	 * Clock operation may occur at runtime and during specific
+	 * system power transition (PM power off, suspend and resume)
+	 * that operate for example in fastcall SMC entries of the PSCI
+	 * services where non-secure world is not operational. In these
+	 * cases we cannot take a mutex and will expect the mutex is
+	 * unlocked.
```

Comment:
I agree it is a bit fragile. I mimicked what's we done for the regulator framework:
https://github.com/OP-TEE/optee_os/blob/4.0.0/core/drivers/regulator/regulator.c#L29-L55

Maybe find a better way to ensure we're in a legitimate PM sequence for allowing not the take a semaphore.
Should I address that now? or find a more elegant solution for both regulator and clock in a seperated P-R?

Reply:
How about taking a spinlock instead when `thread_get_id_may_fail() == THREAD_ID_INVALID`? If normal world is behaving it shouldn't be a problem.

Reply:
You mean taking a spinlock without masking interrupts? Hmm... I did not though about it. I'll consider, thanks.

Reply:
Maybe something like:
```c
static void lock_clk(void)
{
	if (thread_get_id_may_fail() == THREAD_ID_INVALID) {
		if (!spinlock_trylock(&clk_lock) || clk_mutex.state)
			panic();
	} else {
		mutex_lock(&clk_mutex))
		if (clk_lock)
			panic();
	}
}

static void unlock_clk(void)
{
	if (thread_get_id_may_fail() == THREAD_ID_INVALID) {
		if (clk_mutex.state)
			panic();
		spinlock_unlock(&clk_lock);
	} else {
		mutex_unlock(&clk_mutex))
	}
}

Reply:
Yes, something like that. I think it could be improved further as:
``` C
static void lock_clk(void)
{
        if (thread_get_id_may_fail() == THREAD_ID_INVALID)
                assert(!clk_mutex.state);
        else    
                mutex_lock(&clk_mutex);
        if (!cpu_spin_trylock(&clk_lock))
                panic();
}

static void unlock_clk(void)
{       
        if (thread_get_id_may_fail() == THREAD_ID_INVALID)
                assert(!clk_mutex.state);
        else
                mutex_unlock(&clk_mutex);
        cpu_spin_unlock(&clk_lock);
}
```

Reply:
Hmm, I realize that `cpu_spin_trylock()` may not like that we try to lock without exceptions masked. I guess we could mask them temporarily during the call.